transient/config.py

```diff
@@ -0,0 +1,261 @@
+    qmp_timeout = fields.Int(allow_none=True)
+    shutdown_timeout = fields.Int(allow_none=True)
+    ssh_command = fields.Str(allow_none=True)
+    ssh_bin_name = fields.Str(allow_none=True)
+    ssh_port = fields.Int(allow_none=True)
+    ssh_timeout = fields.Int(allow_none=True)
+    ssh_user = fields.Str(allow_none=True)
+    ssh_console = fields.Bool(allow_none=True)
+    ssh_with_serial = fields.Bool(allow_none=True)
+    shared_folder = fields.List(fields.Str(), allow_none=True)
+
+    # marshmallow's decorator post_load() is untyped, forcing create_args()
+    # to be untyped. Therefore, we ignore it to silence the type checker
+    @post_load  # type: ignore
+    def create_config(self, data: Dict[Any, Any], **kwargs: Dict[Any, Any]) -> Config:
+        """Returns the Config dictionary after a schema is loaded and validated
+        """
+        return Config(**data)
+
+
+def replace_hyphens_with_underscores_in_dict_keys(
+    dictionary: Dict[Any, Any]
+) -> Dict[Any, Any]:
+    """Replaces hyphens in the dictionary keys with underscores
+
+       This is the expected key format for TransientConfigSchema
+    """
+    final_dict = {}
+    for k, v in dictionary.items():
+        # Perform this method recursively for sub-directories
+        if isinstance(dictionary[k], dict):
+            value = replace_hyphens_with_underscores_in_dict_keys(v)
+            final_dict[k.replace("-", "_")] = value
+        else:
+            final_dict[k.replace("-", "_")] = v
+
+    return final_dict
+
+
+def parse_config_file(config_file_path: str) -> MutableMapping[str, Any]:
+    """Parses the given config file and returns the contents as a dictionary
+    """
+    with open(config_file_path) as file:
+        config_file = file.read()
+
+    try:
+        parsed_config_file = toml.loads(config_file)
+    except RuntimeError as error:
+        print(error, file=sys.stderr)
+        sys.exit(1)
+
+    return parsed_config_file
+
+
+def get_line_number_of_string_in_config_file(string: str, config_file_path: str) -> int:
+    """Returns the line number where the config file contains the given string
+    """
+    with open(config_file_path) as config_file:
+        for line_number, line in enumerate(config_file, start=1):
+            if string in line:
+                return line_number
+
+    print("Line not found", file=sys.stderr)
+
+    return -1
+
+
+def print_error_message_for_bad_config_file(
+    error: ValidationError, config_file_path: str
+) -> None:
+    """Prints an error message regarding a bad configuration in the config file
+       and its associated line number
+    """
+    for key in error.messages:
+
+        # Revert the key to its preformatted state
+        key = key.replace("_", "-")
+
+        line_number = get_line_number_of_string_in_config_file(key, config_file_path)
+        print(f"Invalid configuration file, line {line_number}: {error}", file=sys.stderr)
+
+
+def load_config_file(config_file_path: str) -> Config:
```

Comment:
So this is the public function? Then `parse_config_file` should be prefixed with a "_".

Reply:
Technically, the only function that's invoked by an external object is `create_transient_run_config_from_cli_args()`

Therefore, I have prefixed all other functions in this class with the underscore prefix.